pkg/wing/apiserver/apiserver.go

```diff
@@ -55,29 +55,31 @@ type WingServer struct {
 }
 
 type completedConfig struct {
-	*Config
+	GenericConfig genericapiserver.CompletedConfig
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
 }
 
 // Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
-func (c *Config) Complete() completedConfig {
-	c.GenericConfig.Complete()
+func (cfg *Config) Complete() CompletedConfig {
+	c := completedConfig{
+		cfg.GenericConfig.Complete(),
+	}
 
 	c.GenericConfig.Version = &version.Info{
 		Major: "1",
 		Minor: "0",
 	}
 
-	return completedConfig{c}
-}
-
-// SkipComplete provides a way to construct a server instance without config completion.
-func (c *Config) SkipComplete() completedConfig {
-	return completedConfig{c}
+	return CompletedConfig{&c}
 }
 
 // New returns a new instance of WingServer from the given config.
 func (c completedConfig) New() (*WingServer, error) {
-	genericServer, err := c.Config.GenericConfig.SkipComplete().New("wing", genericapiserver.EmptyDelegate) // completion is done in Complete, no need for a second time
+	genericServer, err := c.GenericConfig.New("wing", genericapiserver.EmptyDelegate)
```

Comment:
Not using 'SkipComplete' looks good to me, as we already store a reference to the completed config on L68 (https://github.com/jetstack/tarmak/pull/15/commits/40390ed7db67156ea79b4d3a9a0fe30b738d377a#diff-eeee2947c8ccb06990948a4196759b69R68).

In the previous code we threw away the `completedConfig` structure returned from `cfg.GenericConfig.Complete`, which meant we had to re-construct this completed configuration (albeit this time without calling Complete - I assume this was okay because Complete was originally called on a pointer? Not sure as I've not checked the code for `Complete` yet.

Reply:
Just wanted someone to double check